File: examples/python/food_image_recognize_stream.py
#!/usr/bin/env python3
"""流式图片识别示例

用法:
  python food_image_recognize_stream.py                              # 默认图片
  python food_image_recognize_stream.py --url https://xxx.jpg        # 按URL识别
  python food_image_recognize_stream.py --base64 /path/to/image.jpg  # 按base64识别
  python food_image_recognize_stream.py --url https://xxx.jpg --ingredients sugar,purine --with-nrv
"""
import os
import sys
import json
import base64

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../python'))

from boohee_sdk import BooheeClient, AuthMode, BaseReq, HttpMethod


class FoodImageRecognizeStreamReq(BaseReq):
    def __init__(self, image_url=None, image_data=None, ingredients=None, with_nrv=False):
        self.image_url = image_url
        self.image_data = image_data
        self.ingredients = ingredients
        self.with_nrv = with_nrv

    def get_method(self) -> HttpMethod:
        return HttpMethod.POST

    def get_url(self) -> str:
        return '/open-apis/v1/food/image_recongize_stream'

    def get_body(self):
        body = {}
        if self.image_url:
            body['image_url'] = self.image_url
        if self.image_data:
            body['image_data'] = self.image_data
        if self.ingredients:
            body['ingredients'] = self.ingredients
        if self.with_nrv:
            body['with_nrv'] = True
        return body


def load_env():
    env_path = os.path.join(os.path.dirname(__file__), '../../.env')
    if not os.path.exists(env_path):
        print("请先复制 .env.example 为 .env 并填入真实凭证")
        sys.exit(1)
    with open(env_path) as f:
        for line in f:
            line = line.strip()
            if line and not line.startswith('#') and '=' in line:
                key, value = line.split('=', 1)
                os.environ.setdefault(key.strip(), value.strip())


def parse_args():
    args = sys.argv[1:]
    opts = {}
    i = 0
    while i < len(args):
        if args[i] == '--url':
            opts['image_url'] = args[i + 1]; i += 2
        elif args[i] == '--base64':
            filepath = args[i + 1]
            with open(filepath, 'rb') as f:
                opts['image_data'] = base64.b64encode(f.read()).decode()
            i += 2
        elif args[i] == '--ingredients':
            opts['ingredients'] = args[i + 1]; i += 2
        elif args[i] == '--with-nrv':
            opts['with_nrv'] = True; i += 1
        else:
            i += 1
    return opts


def main():
    load_env()

    api_key = os.environ.get('BOOHEE_API_KEY')
    if not api_key:
        print("BOOHEE_API_KEY 未配置")
        sys.exit(1)

    client = BooheeClient(
        api_key=api_key,
        auth_mode=AuthMode.API_KEY,
        base_url=os.environ.get('BOOHEE_BASE_URL', 'https://api.boohee.com'),
    )
    opts = parse_args()

    if not opts.get('image_url') and not opts.get('image_data'):
        opts['image_url'] = 'https://img.boohee.cn/ghp/foods/000/000/001/140_1.jpg'

    label = opts.get('image_url') or 'base64数据'
    print(f"识别图片: {label}\n")
    print("流式输出:")

    full_content = ''
    food_info = []

    for chunk in client.execute_stream(FoodImageRecognizeStreamReq(**opts)):
        try:
            event = json.loads(chunk)
            result_code = event.get('result_code', -1)
            if result_code != 0:
                print(f"\n错误: {event.get('message', 'unknown')}")
                sys.exit(1)

            data = event.get('data', {})
            content = data.get('content', '')
            if content:
                print(content, end='', flush=True)
                full_content += content

            if data.get('end'):
                food_info = data.get('food_info', [])
        except json.JSONDecodeError:
            pass

    print("\n")

    if food_info:
        print(f"--- 识别到 {len(food_info)} 个食物 ---\n")
        for food in food_info:
            print(f"  {food.get('name', '?')}  "
                  f"估重: {food.get('amount', '?')}g  "
                  f"热量: {food.get('calories', '?')}kcal/100g  "
                  f"蛋白质: {food.get('protein', '?')}g  "
                  f"脂肪: {food.get('fat', '?')}g  "
                  f"碳水: {food.get('carbohydrate', '?')}g")
            ingredients = food.get('ingredients', [])
            if ingredients:
                ing_strs = [f"{i['name']}: {i['value']}{i['unit']}" for i in ingredients]
                print(f"    营养素: {', '.join(ing_strs)}")
            nrv = food.get('nrv', [])
            if nrv:
                nrv_strs = [f"{n['name']}: {n['value']}" for n in nrv]
                print(f"    NRV: {', '.join(nrv_strs)}")
            print()


if __name__ == '__main__':
    main()

File: examples/nodejs/food_image_recognize_stream.ts
/**
 * 流式图片识别示例
 *
 * 用法:
 *   npx tsx food_image_recognize_stream.ts                              # 默认图片
 *   npx tsx food_image_recognize_stream.ts --url https://xxx.jpg        # 按URL识别
 *   npx tsx food_image_recognize_stream.ts --base64 /path/to/image.jpg  # 按base64识别
 *   npx tsx food_image_recognize_stream.ts --url https://xxx.jpg --ingredients sugar,purine --with-nrv
 */
import * as fs from 'fs';
import * as path from 'path';
import { BooheeClient, AuthMode, HttpMethod, Request } from '../../nodejs/src';

class FoodImageRecognizeStreamReq implements Request {
  private body: Record<string, any>;

  constructor(opts: {
    imageUrl?: string; imageData?: string;
    ingredients?: string; withNrv?: boolean;
  }) {
    this.body = {};
    if (opts.imageUrl) this.body.image_url = opts.imageUrl;
    if (opts.imageData) this.body.image_data = opts.imageData;
    if (opts.ingredients) this.body.ingredients = opts.ingredients;
    if (opts.withNrv) this.body.with_nrv = true;
  }

  getMethod() { return HttpMethod.POST; }
  getUrl() { return '/open-apis/v1/food/image_recongize_stream'; }
  getBody() { return this.body; }
}

function loadEnv(): void {
  const envPath = path.resolve(__dirname, '../../.env');
  if (!fs.existsSync(envPath)) {
    console.log('请先复制 .env.example 为 .env 并填入真实凭证');
    process.exit(1);
  }
  const content = fs.readFileSync(envPath, 'utf-8');
  for (const line of content.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const idx = trimmed.indexOf('=');
    if (idx > 0) {
      const key = trimmed.substring(0, idx).trim();
      const value = trimmed.substring(idx + 1).trim();
      if (!process.env[key]) {
        process.env[key] = value;
      }
    }
  }
}

function parseArgs(): {
  imageUrl?: string; imageData?: string;
  ingredients?: string; withNrv: boolean;
} {
  const opts = { withNrv: false } as any;
  const args = process.argv.slice(2);
  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '--url': opts.imageUrl = args[++i]; break;
      case '--base64':
        const fileBytes = fs.readFileSync(args[++i]);
        opts.imageData = fileBytes.toString('base64');
        break;
      case '--ingredients': opts.ingredients = args[++i]; break;
      case '--with-nrv': opts.withNrv = true; break;
    }
  }
  return opts;
}

async function main(): Promise<void> {
  loadEnv();

  const apiKey = process.env.BOOHEE_API_KEY;
  if (!apiKey) {
    console.log('BOOHEE_API_KEY 未配置');
    process.exit(1);
  }

  const client = new BooheeClient({
    authMode: AuthMode.API_KEY,
    apiKey,
    baseUrl: process.env.BOOHEE_BASE_URL || 'https://api.boohee.com',
  });

  const opts = parseArgs();
  if (!opts.imageUrl && !opts.imageData) {
    opts.imageUrl = 'https://img.boohee.cn/ghp/foods/000/000/001/140_1.jpg';
  }

  const label = opts.imageUrl || 'base64数据';
  console.log(`识别图片: ${label}\n`);
  console.log('流式输出:');

  let fullContent = '';
  let foodInfo: any[] = [];

  for await (const chunk of client.executeStream(new FoodImageRecognizeStreamReq(opts))) {
    try {
      const event = JSON.parse(chunk);
      if (event.result_code !== 0) {
        console.log(`\n错误: ${event.message}`);
        process.exit(1);
      }
      const data = event.data ?? {};
      if (data.content) {
        process.stdout.write(data.content);
        fullContent += data.content;
      }
      if (data.end) {
        foodInfo = data.food_info ?? [];
      }
    } catch {
      // 忽略解析错误
    }
  }

  console.log();

  if (foodInfo.length > 0) {
    console.log(`\n--- 识别到 ${foodInfo.length} 个食物 ---\n`);
    for (const food of foodInfo) {
      console.log(`  ${food.name ?? '?'}  估重: ${food.amount ?? '?'}g  ` +
        `热量: ${food.calories ?? '?'}kcal/100g  ` +
        `蛋白质: ${food.protein ?? '?'}g  ` +
        `脂肪: ${food.fat ?? '?'}g  ` +
        `碳水: ${food.carbohydrate ?? '?'}g`);
      if (food.ingredients?.length) {
        const ingStrs = food.ingredients.map(
          (i: any) => `${i.name}: ${i.value}${i.unit}`);
        console.log(`    营养素: ${ingStrs.join(', ')}`);
      }
      if (food.nrv?.length) {
        const nrvStrs = food.nrv.map(
          (n: any) => `${n.name}: ${n.value}`);
        console.log(`    NRV: ${nrvStrs.join(', ')}`);
      }
      console.log();
    }
  }
}

main().catch(console.error);
